src/gui/folderstatusmodel.cpp

```diff
@@ -863,69 +873,68 @@ void FolderStatusModel::slotApplySelectiveSync()
+void FolderStatusModel::slotSetProgress(const ProgressInfo &progress, Folder *f)
 {
-    auto par = qobject_cast<QWidget *>(QObject::parent());
-    if (!par->isVisible()) {
+    if (!qobject_cast<QWidget *>(QObject::parent())->isVisible()) {
         return; // for https://github.com/owncloud/client/issues/2648#issuecomment-71377909
     }
 
-    Folder *f = qobject_cast<Folder *>(sender());
-    if (!f) {
-        return;
-    }
-
-    int folderIndex = -1;
-    for (int i = 0; i < _folders.count(); ++i) {
-        if (_folders.at(i)._folder == f) {
-            folderIndex = i;
-            break;
+    const int folderIndex = [f, this] {
+        int folderIndex = -1;
+        for (int i = 0; i < _folders.count(); ++i) {
+            if (_folders.at(i)._folder == f) {
+                folderIndex = i;
+                break;
+            }
         }
-    }
+        return folderIndex;
+    }();
     if (folderIndex < 0) {
         return;
     }
 
     auto *pi = &_folders[folderIndex]._progress;
 
-    QVector<int> roles;
-    roles << FolderStatusDelegate::SyncProgressItemString
-          << FolderStatusDelegate::WarningCount
-          << Qt::ToolTipRole;
+    const QVector<int> roles = { FolderStatusDelegate::SyncProgressItemString, FolderStatusDelegate::WarningCount, Qt::ToolTipRole };
 
     if (progress.status() == ProgressInfo::Discovery) {
         if (!progress._currentDiscoveredRemoteFolder.isEmpty()) {
             pi->_overallSyncString = tr("Checking for changes in remote '%1'").arg(progress._currentDiscoveredRemoteFolder);
             emit dataChanged(index(folderIndex), index(folderIndex), roles);
-            return;
         } else if (!progress._currentDiscoveredLocalFolder.isEmpty()) {
             pi->_overallSyncString = tr("Checking for changes in local '%1'").arg(progress._currentDiscoveredLocalFolder);
             emit dataChanged(index(folderIndex), index(folderIndex), roles);
-            return;
         }
-    }
-
-    if (progress.status() == ProgressInfo::Reconcile) {
+    } else if (progress.status() == ProgressInfo::Reconcile) {
         pi->_overallSyncString = tr("Reconciling changes");
         emit dataChanged(index(folderIndex), index(folderIndex), roles);
-        return;
-    }
+    } else {
+        // Status is Starting, Propagation or Done
 
-    // Status is Starting, Propagation or Done
+        if (!progress._lastCompletedItem.isEmpty()
+            && Progress::isWarningKind(progress._lastCompletedItem._status)) {
+            pi->_warningCount++;
+        }
 
-    if (!progress._lastCompletedItem.isEmpty()
-        && Progress::isWarningKind(progress._lastCompletedItem._status)) {
-        pi->_warningCount++;
+        // progress updates are expensive, throtle them
+        if (std::chrono::system_clock::now() - _lastProgressUpdated > progressUpdateTimeOutC) {
+            computeProgress(progress, pi);
+            _lastProgressUpdated = std::chrono::system_clock::now();
```

Comment:
Ah, there is this QElapsedTimer which is very nice and neat...

Reply:
Yes, progressUpdateTimeOutC would be a qint64, which would result in more clear code for me. But... yes.